shared-libs/cht-datasource/src/index.ts

```diff
@@ -0,0 +1,45 @@
+/**
+ * CHT Script API - Index
+ * Builds and exports a versioned API from feature modules.
+ * Whenever possible keep this file clean by defining new features in modules.
+ */
+import { hasAnyPermission, hasPermissions } from './auth';
+import { V1 as PersonV1 } from './person';
+import { V1 as DocV1 } from './libs/doc';
+import { Nullable as _Nullable } from './libs/core';
+import getLocalEnvironment, { LocalEnvironment, SourceDatabases as _SourceDatabases } from './libs/local-environment';
+
+const adapt = <T>(
+  localEnv: Nullable<LocalEnvironment>,
+  remoteAdapter: T,
+  localAdapter: (localEnv: LocalEnvironment) => T,
+) => {
+  if (localEnv) {
+    return localAdapter(localEnv);
+  }
+  return remoteAdapter;
+};
+
+export type SourceDatabases = _SourceDatabases;
+export type Nullable<T> = _Nullable<T>;
+
+export namespace V1 {
+  export import Person = PersonV1.Person;
+  export import UuidIdentifier = DocV1.UuidIdentifier;
+  export import byUuid = DocV1.byUuid;
+}
+
+const getDataSource = async (sourceDatabases?: SourceDatabases) => {
+  const localEnv = await getLocalEnvironment(sourceDatabases);
+
+  return {
+    v1: {
+      hasPermissions,
+      hasAnyPermission,
+      person: adapt(localEnv, PersonV1.remote, PersonV1.local),
```

Comment:
I'm slightly on the fence on this one. I think there are two options...
1. what you've done - Person knows how to get a person from local or remote dbs, or
2. what pouchdb does - local and remote are datasource adapters that implement the same interface (getPerson)

This distinction doesn't change the external API at all - this is all internal to the datasource lib.

I prefer option 2 personally. I think it's a better separation of concerns, ie: Person knows about people, the remote datasource cares about http, the local datasource cares about pouch.

It feels a little dirty because it's a second layer of abstraction but on balance I think it's cleaner.

Reply:
So, funny story, I was originally trying to avoid #2 for a couple reasons:

- I thought maybe I would push the local/remote abstraction down so it was internal to `person.ts`
- I did not want to have to maintain a common interface

As I iterated on this code, I decided it seemed better to make the local/remote distinction at the highest level possible and not encapsulate it in `person.ts`. Then I also ended up with a common interface anyway since that was the best way to keep everything in sync.  

Thinking about this all again, I totally agree with you that the code will be a lot cleaner if we separate local/remote into different files (and maybe different directories)....  In reality, they will not be sharing code anyway.